File: app/lib/models/company.dart
import 'package:json_annotation/json_annotation.dart';

part 'company.g.dart';

@JsonSerializable()
class Company {
  final String id;
  final String name;
  final Map<String, dynamic> settings;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isActive;

  Company({
    required this.id,
    required this.name,
    Map<String, dynamic>? settings,
    required this.createdAt,
    required this.updatedAt,
    this.isActive = true,
  }) : settings = settings ?? {};

  factory Company.fromJson(Map<String, dynamic> json) => _$CompanyFromJson(json);
  Map<String, dynamic> toJson() => _$CompanyToJson(this);

  Company copyWith({
    String? id,
    String? name,
    Map<String, dynamic>? settings,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isActive,
  }) {
    return Company(
      id: id ?? this.id,
      name: name ?? this.name,
      settings: settings ?? this.settings,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isActive: isActive ?? this.isActive,
    );
  }

  bool isValidName() {
    return name.isNotEmpty && name.length <= 100;
  }

  T? getSetting<T>(String key, {T? defaultValue}) {
    if (settings.containsKey(key)) {
      return settings[key] as T?;
    }
    return defaultValue;
  }

  Company updateSetting(String key, dynamic value) {
    final newSettings = Map<String, dynamic>.from(settings);
    newSettings[key] = value;
    return copyWith(settings: newSettings);
  }

  Company removeSetting(String key) {
    final newSettings = Map<String, dynamic>.from(settings);
    newSettings.remove(key);
    return copyWith(settings: newSettings);
  }

  bool get syncEnabled => getSetting<bool>('syncEnabled', defaultValue: true) ?? true;
  int get syncIntervalMinutes => getSetting<int>('syncIntervalMinutes', defaultValue: 30) ?? 30;
  int get maxPhotosPerSync => getSetting<int>('maxPhotosPerSync', defaultValue: 100) ?? 100;
  bool get autoAssignByGPS => getSetting<bool>('autoAssignByGPS', defaultValue: true) ?? true;
  int get photoQuality => getSetting<int>('photoQuality', defaultValue: 95) ?? 95;
  bool get keepOriginalPhotos => getSetting<bool>('keepOriginalPhotos', defaultValue: true) ?? true;
  String get defaultPhotoNamingPattern => 
      getSetting<String>('defaultPhotoNamingPattern', defaultValue: '{equipment}_{timestamp}') ?? 
      '{equipment}_{timestamp}';
}

File: api/src/models/company.js
const { DataTypes } = require('sequelize');
const sequelize = require('../database/connection');

const Company = sequelize.define('Company', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  name: {
    type: DataTypes.STRING(100),
    allowNull: false,
    unique: true,
    validate: {
      notEmpty: true,
      len: [1, 100],
    },
  },
  settings: {
    type: DataTypes.JSON,
    defaultValue: {},
  },
  createdAt: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  updatedAt: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  isActive: {
    type: DataTypes.BOOLEAN,
    defaultValue: true,
  },
}, {
  tableName: 'companies',
  timestamps: true,
  indexes: [
    {
      fields: ['name'],
      unique: true,
    },
    {
      fields: ['isActive'],
    },
  ],
});

Company.prototype.toJSON = function() {
  const values = { ...this.get() };
  return values;
};

Company.prototype.getSetting = function(key, defaultValue = null) {
  return this.settings[key] !== undefined ? this.settings[key] : defaultValue;
};

Company.prototype.updateSetting = async function(key, value) {
  this.settings = {
    ...this.settings,
    [key]: value,
  };
  await this.save();
};

Company.prototype.removeSetting = async function(key) {
  const newSettings = { ...this.settings };
  delete newSettings[key];
  this.settings = newSettings;
  await this.save();
};

Company.prototype.getDefaultSettings = function() {
  return {
    syncEnabled: this.getSetting('syncEnabled', true),
    syncIntervalMinutes: this.getSetting('syncIntervalMinutes', 30),
    maxPhotosPerSync: this.getSetting('maxPhotosPerSync', 100),
    autoAssignByGPS: this.getSetting('autoAssignByGPS', true),
    photoQuality: this.getSetting('photoQuality', 95),
    keepOriginalPhotos: this.getSetting('keepOriginalPhotos', true),
    defaultPhotoNamingPattern: this.getSetting('defaultPhotoNamingPattern', '{equipment}_{timestamp}'),
    maxStorageGB: this.getSetting('maxStorageGB', 100),
    dataRetentionDays: this.getSetting('dataRetentionDays', null),
    allowOfflineMode: this.getSetting('allowOfflineMode', true),
  };
};

Company.findActive = function() {
  return this.findAll({
    where: { isActive: true },
    order: [['name', 'ASC']],
  });
};

Company.findByName = function(name) {
  return this.findOne({
    where: { name },
  });
};

Company.associate = function(models) {
  Company.hasMany(models.Client, {
    foreignKey: 'companyId',
    as: 'clients',
  });
  Company.hasMany(models.User, {
    foreignKey: 'companyId',
    as: 'users',
  });
};

module.exports = Company;